### ISSUE TITLE
mp3 SoundSource elapsed() is incorrect (increments too fast)

### ISSUE THREAD

**Issue by tarkah** [ggez/ggez#632]
**Describe the bug**
SoundSource created from an mp3 file has accurate audio playback, however the duration returned from elapsed() is inaccurate.  The duration increments much faster than the actual playback.  This problem does not occur with ogg. Issue occurs on both windows and linux.

**To Reproduce**
Add any mp3 file to the resources folder and edit `examples/sounds` to load that mp3 file.

Run `cargo run --example sounds` and notice `Elapsed time` increments faster than audio playback.

**Expected behavior**
`Elapsed time: seconds` should line up with audio playback.

**Screenshots**
![mintty_2019-05-16_15-48-11](https://user-images.githubusercontent.com/10239377/57892115-a5440480-77f2-11e9-96e8-6e8a0198408b.png)

**Hardware and Software:**
 - ggez version: 0.5.2-r.2 & current git master
 - OS: Windows 10 and Linux (Solus)
 - Graphics card: N/A
 - Graphics card drivers: N/A

**Comment by tarkah** [comment]
#454 Implements this.  @mbillingr do you know how to fix this?

**Comment by tarkah** [comment]
I've tried the following example directly against the Rodio crate v0.8.1 and it seems to increment just fine... I don't see much difference between the below and how the periodic access used in ggez.

    extern crate rodio;
    
    use std::io::BufReader;
    use rodio::Source;
    use std::sync::atomic::{AtomicUsize, Ordering};
    
    fn main() {
        let device = rodio::default_output_device().unwrap();
        let sink = rodio::Sink::new(&device);
    
        let file = std::fs::File::open("test.mp3").unwrap();
        let counter = AtomicUsize::new(0);
        let source = rodio::Decoder::new(BufReader::new(file)).unwrap().periodic_access(std::time::Duration::from_millis(1000), move |s| {
            let count = counter.fetch_add(1000, Ordering::SeqCst);
            println!("1 sec passed, sample rate {:?}, count {:?}", s.sample_rate(), count);

        });
        sink.append(source);
        sink.sleep_until_end();
    }

**Comment by mbillingr** [comment]
Does the mp3 have variable sample rate?

There is this comment in [Rodio's periodic access implementation](https://github.com/tomaka/rodio/blob/master/src/source/periodic.rs#L12) that says
 ```
    // TODO: handle the fact that the samples rate can change
    // TODO: generally, just wrong
    let update_ms = period.as_secs() as u32 * 1_000 + period.subsec_nanos() / 1_000_000;
    let sample_rate = source.sample_rate() * source.channels() as u32;
    let update_frequency = (update_ms * sample_rate) / 1000;
```

They set the update interval during initialization based on the sample rate. If the sample rate changes during playback the update interval will be wrong.
I'm really not an expert on audio file formats. Does mp3 have a dynamic sample rate? I'm sure it has dynamic bit rate, but could this affect the sample rate - on purpose or due to a bug?

**Comment by tarkah** [comment]
@mbillingr I am not sure.  However, I tested `periodic_access` using just the Rodio crate as per my comment above and didn't have any issues, using the exact same mp3 file. That code isn't much different than how it's implemented in ggez... so I am really at a loss.

**Comment by tarkah** [comment]
@mbillingr Check this out: [link](https://github.com/tarkah/ggez/blob/mp3-sample-rate/mp3_sample_rate/src/main.rs)

I created a branch to debug this thing, and it appears to be caused when AtomicUsize is Arc'd and loaded in a different thread.  Inside the closure, it displays the correct time. Wish I had better knowledge to know why this is happening.

![Screenshot from 2019-05-16 22-53-28](https://user-images.githubusercontent.com/10239377/57906324-692e9500-782e-11e9-9cdf-b855ad3f7f8b.png)

**Comment by mbillingr** [comment]
Agreed, it's strange. There is no obvious difference between ggez and your test code.

I'd like to look into the issue in detail. Can you share some of the mp3 files you use?

**Comment by icefoxen** [comment]
Thanks for the interesting bug report!  Probably-useless-but-paranoid suggestions:

 * crank down the query interval in your example to match ggez's 100 ms default
 * Use `detach()` instead of `sleep_until_end()` just in case there's some weird threading stuff happening
 * Similarly, send your debug message over an mpsc channel and print it outside the closure, `println!()` involves locking a mutex on stdout.

I don't expect any of these to change anything but we can at least rule them out.  I'd also love to take a look at the mp3 files themselves.

**Comment by tarkah** [comment]
@icefoxen @mbillingr It'd help if I took a look at the commit history for Rodio.  [https://github.com/tomaka/rodio/commit/dba627a1b4b0562b6b30b17befc3108e356f3e6c](https://github.com/tomaka/rodio/commit/dba627a1b4b0562b6b30b17befc3108e356f3e6c) resolves this.  No new release has been created since this commit.

**Comment by icefoxen** [comment]
Good find, thanks!  Leave this issue open so we know to double-check it and update rodio when possible.

**Comment by tarkah** [comment]
Will do. Thanks! I can't wait to switch over my project from using SFML, I can't stand having to deal with it as a dependency.

**Comment by mbillingr** [comment]
@tarkah  Sorry I am slow on the upbeat here. How come your isolated test worked correctly if the fix in Rodio has not been released yet?

**Comment by tarkah** [comment]
@mbillingr I didn't pay attention to when v0.8.1 was released and I actually tested off cloned master yesterday, which has it fixed. I happened to try again this morning on a different computer, but instead of cloning Rodio master, I used the v0.8.1 crate. Being that the test no longer worked, it immediately led me to the conclusion that Rodio was patched somewhere between the release ggez uses and master.

Lo and behold, there was a commit to fix this exact problem! Testing right before the commit and mp3 (2 channels) increments too fast. Testing the commit and it increments just fine! So you were basically spot on with your suggestion, it had to do with the sample rate, being that sample rate needed to be multiplied by # channels.

**Comment by mbillingr** [comment]
@tarkah allright. Thank you for the patient explanation :) I actually saw the commit yesterday but did not think twice about it. So good job in reaching the correct conclusion!

**Comment by tarkah** [comment]
@mbillingr Of course, thanks for the help!

**Comment by FireCubez** [comment]
Is this fixed? I'm using `0.5.0-rc.2` and having the same issue.

**Comment by icefoxen** [comment]
The `devel` branch has been updated to`rodio` 0.9, give that a try and see if it's still a problem!
